### PR TITLE
use new API for ens name registration

### LINKED ISSUE #12891: Synchronize ENS name across devices
As a user
I want my ENS name to be synchronized across paired devices
So that I don't have to set it up on each one of them

### Description

When setting up an ENS name on one device, the change should also be reflected on each paired device you have.

Ideally this is also backed up similarly as we do with contacts, but this can be implemented separately.

It can be tested on Ropsten so you don't have to spend real ETH, you can ask the QA team for assistance with the faucet. 

Most of the changes (all of them probably), are in status-go.

### PR DESCRIPTION
**Pull request by qfrank**
this PR mainly focused on following requirements:
- use new APIs for ens name operations which are also used by desktop
- use ens names from table `ens_usernames` rather than `settings.usernames` anymore
- allow synchronize ENS name across devices
- support ENS name backup

fixes #12891

#### Platforms

- Android
- iOS

status: ready

### CODE REVIEW

File: src/status_im/ens/core.cljs
Comment (Parveshdhull) [inline src/status_im/ens/core.cljs]: Please move closing bracket to above line

File: src/status_im2/navigation/core.cljs
Comment (Parveshdhull) [inline src/status_im2/navigation/core.cljs]: Please revert these changes, not needed anymore

File: src/status_im/ens/core.cljs
Comment (Parveshdhull) [inline src/status_im/ens/core.cljs]: should we revert it?

File: src/status_im2/subs/ens.cljs
Comment (flexsurfer) [inline src/status_im2/subs/ens.cljs]: i'm not sure why `in-progress-names` is needed, why not just get names in reduce?

File: src/status_im2/navigation/core.cljs
Comment (flexsurfer) [inline src/status_im2/navigation/core.cljs]: why is it commented and not removed?

File: src/status_im2/config.cljs
Comment (flexsurfer) [inline src/status_im2/config.cljs]: probably this is leftover after development? it shouldn't be enabled by default

File: src/status_im/ethereum/transactions/core.cljs
Comment (flexsurfer) [inline src/status_im/ethereum/transactions/core.cljs:262]: concat is not needed anymore

File: src/status_im/signing/core.cljs
Comment (flexsurfer) [inline src/status_im/signing/core.cljs]: better to use keys

File: src/status_im2/config.cljs
Comment (qfrank) [reply src/status_im2/config.cljs]: I was wondering why we didn't enable it in development env, can i add `TEST_STATEOFUS=1` into `.env` so it enabled in dev env (i'm not sure) ?  @flexsurfer

File: src/status_im2/subs/ens.cljs
Comment (qfrank) [reply src/status_im2/subs/ens.cljs]: because `:ens/names` contains names that corresponding transaction may be still pending state, but transaction of names returned from `:ens.main/screen` need to be completed.

Comment (flexsurfer) [reply src/status_im2/subs/ens.cljs]: i mean `in-progress-names` is used only once, so why do we need it? why don't have `not-in-progress-names` ? so we don't need `(remove #(contains?`

File: src/status_im2/config.cljs
Comment (flexsurfer) [reply src/status_im2/config.cljs]: yes sure

File: src/status_im/ethereum/transactions/core.cljs
Comment (qfrank) [reply src/status_im/ethereum/transactions/core.cljs:262]: emm.. if replace `(concat [] (mapv add-transfer transfers))` to `(mapv add-transfer transfers)` will get error:
```
 ERROR  Error: function status_im$ethereum$transactions$core$add_transfer(var_args){
var G__86827 = arguments.length;
switch (G__86827) {
case 1:
return status_im.ethereum.transactions.core.add_transfer.cljs$core$IFn$_invoke$arity$1((arguments[(0)]));

break;
case 2:
return status_im.ethereum.transactions.core.add_transfer.cljs$core$IFn$_invoke$arity$2((arguments[(0)]),(arguments[(1)]));

break;
default:
throw (new Error(["Invalid arity: ",cljs.core.str.cljs$core$IFn$_invoke$arity$1(arguments.length)].join('')));

}
} is not ISeqable
``` 

@flexsurfer

Comment (flexsurfer) [reply src/status_im/ethereum/transactions/core.cljs:262]: that's strange

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/ethereum/transactions/core.cljs
Comment (qfrank) [reply src/status_im/ethereum/transactions/core.cljs:262]: 🤷‍♂️

**Review by Parveshdhull — APPROVED** [review]

**Review by yqrashawn — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
Hi @qfrank could you please rebase the PR? Thanx

**Comment by qfrank** [comment]
> Hi @qfrank could you please rebase the PR? Thanx

done! @VolodLytvynenko

**Comment by churik** [comment]
@qfrank 

we're close to release now, I'd prefer to postpone this PR until the branch cut.

**Comment by churik** [comment]
@qfrank 
I've tried to buy ENS name on goerli on your PR and the issue is that after purchasing it is not set in profile automatically and there is no other option to set it.


How it behaves in current `develop` after purchase:


https://github.com/status-im/status-mobile/assets/4557972/b3e5fc31-a8f5-45d5-a779-5246bfdf85bc

How it behaves in PR after purchase:

https://github.com/status-im/status-mobile/assets/4557972/d466d2d7-0497-4762-974b-c00f0b1c58db

*Steps to reproduce*:
- install PR build
- recover user with ETH / STT on account
- try to buy new ENS

**Comment by qfrank** [comment]
> @qfrank I've tried to buy ENS name on goerli on your PR and the issue is that after purchasing it is not set in profile automatically and there is no other option to set it.
> 
> How it behaves in current `develop` after purchase:
> 
>  FILE.2023-09-05.12.21.14.mp4 
> How it behaves in PR after purchase:
> 
>  FILE.2023-09-05.12.24.00.mp4 
> _Steps to reproduce_:
> 
> * install PR build
> * recover user with ETH / STT on account
> * try to buy new ENS

thank you for your feedback ❤️, this issue should be fixed with last commit 🙏 , note: you need wait registration's state changed to be `finished`(need wait for seconds), once it get finished, it will set in profile automatically @churik

**Comment by pavloburykh** [comment]
@qfrank hi! could you please rebase PR (both go and mobile branches) before we start re-testing? Thank you!

**Comment by qfrank** [comment]
> @qfrank hi! could you please rebase PR (both go and mobile branches) before we start re-testing? Thank you!

done! @pavloburykh

**Comment by pavloburykh** [comment]
@qfrank could you please clarify if ens backing up is implemented in current PR? Or it will be implemented separately?

**Comment by qfrank** [comment]
> @qfrank could you please clarify if ens backing up is implemented in current PR? Or it will be implemented separately?

ens backing up is implemented in current PR @pavloburykh

**Comment by pavloburykh** [comment]
> ens backing up is implemented in current PR @pavloburykh

@qfrank please take a look at couple issues related to ens backup 

### ISSUE 1 Backed up ens is not displayed until first relogin after account restore

**Steps:** 

1. Device 1 creates Account 1
2. Device 1 adds ENS for Account 1
3. Device 1 performs backup
4. Device 2 to restores Account 1
5. See if ENS name is displayed on Device 2 

**Actual result:** ENS is not displayed until first re-login

**Comment by pavloburykh** [comment]
### ISSUE 2 Backed up ens is not applied on login screen account restore

**Steps:** 

1. Device 1 creates Account 1
2. Device 1 adds ENS for Account 1
3. Device 1 performs backup
4. Device 2 to restores Account 1
5. Device 2 peforms relogin and logout
6. See if ENS is applied for Device 2 on login screen 

**Actual result:** ENS is not applied on login screen. Display name is still shown. 

photo_2023-09-13 16 55 07


**Expected result:** ENS name is shown on login screen 

**Video and logs for both ISSUE 1 and ISSUE 2**

[Device 2_logs.zip](https://github.com/status-im/status-mobile/files/12598280/Device.2_logs.zip)

https://github.com/status-im/status-mobile/assets/97245802/f7c36ca0-2026-4093-bd35-0b1999757ceb

**Comment by pavloburykh** [comment]
@cammellos hello!

I have a question regarding ens backing up feature. 

We had a discussion about the above [ISSEU 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638) with @qfrank. In his opinion it might be confusing for user to see backed up ens name applied instead of Display name right after restoring the account, because user has just set his Display name during onboarding. 

I don't see big issue here, but could you still confirm if backing up ens is the feature that want to have now? thank you.

**Comment by cammellos** [comment]
> @cammellos hello!
> 
> I have a question regarding ens backing up feature.
> 
> We had a discussion about the above [ISSEU 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638) with @qfrank. In his opinion it might be confusing for user to see backed up ens name applied instead of Display name right after restoring the account, because user has just set his Display name during onboarding.
> 
> I don't see big issue here, but could you still confirm if backing up ens is the feature that want to have now? thank you.

As discussed we @qfrank 
in the flow we are describing (restore seed phrase -> set your display name)
we should not be fetching backup at all.

so the case we are discussing, in theory, should never happen

but it's just because we don't have the other flow, where you say that "I am an existing user" -> Enter seed phrase
in that flow, we should restore ENS name and the user should see ENS name

I think in this case, the best thing to do, is to act exactly as we would if the user has chosen:

"I am an existing user" -> Enter seed phrase
so we can validate that that flow works fine
and then, once we are happy and we implement that flow UI wise, we know it should work fine, and we just stop fetching backup on the other flow, so that both should work fine

TLDR:
It's fine for the ENS name to override the display name in this case.

**Comment by pavloburykh** [comment]
@cammellos thank you!

@qfrank so, based on

> It's fine for the ENS name to override the display name in this case.

I believe we need to address [ISSUE 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638)

**Comment by qfrank** [comment]
> @cammellos thank you!
> 
> @qfrank so, based on
> 
> > It's fine for the ENS name to override the display name in this case.
> 
> I believe we need to address [ISSUE 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638)

Hi @pavloburykh , issue 2 should be fixed now, pls retry 🙏

**Comment by pavloburykh** [comment]
@qfrank thank you! It seems that [ISSUE 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638) is fixed but partially. 

I still faced display name instead of ENS on login screen. 

Trying to identify the steps. Meanwhile, you can check logs, maybe they will help to figure out the issue. 

[logs (8).zip](https://github.com/status-im/status-mobile/files/12609644/logs.8.zip)

https://github.com/status-im/status-mobile/assets/97245802/d1cdc6f7-6b68-4d79-aa45-b879d7f69663

**Comment by pavloburykh** [comment]
> I still faced display name instead of ENS on login screen. 

@qfrank am not able reproduce it again and identify the steps. Let's ignore it for now. Looks like some edge case. 

Overall PR looks good. 

Could you please rebase the PR (both go and mobile branches) in order to include the fix which unblocks e2e. After that I will trigger and e2e run and we finally will be ready to merge this PR. 

Thank you!

**Comment by qfrank** [comment]
> > I still faced display name instead of ENS on login screen.
> 
> @qfrank am not able reproduce it again and identify the steps. Let's ignore it for now. Looks like some edge case.
> 
> Overall PR looks good.
> 
> Could you please rebase the PR (both go and mobile branches) in order to include the fix which unblocks e2e. After that I will trigger and e2e run and we finally will be ready to merge this PR.
> 
> Thank you!

After rebase for status-go, i found someone else brought new bug 🤦‍♂️: `{"type":"node.crashed","event":{"error":"Key: 'NodeConfig.KeycardPairingDataFile' Error:Field validation for 'KeycardPairingDataFile' failed on the 'required' tag"}}` 

steps to reproduce:
- recover account using seedphrase
- logout
- try to login and you will find you can't 

@pavloburykh 

I'll try to fix this

**Comment by pavloburykh** [comment]
> I'll try to fix this

Thank you @qfrank!

**Comment by qfrank** [comment]
> > I'll try to fix this
> 
> Thank you @qfrank!

done! hope we merge this soon @pavloburykh

**Comment by qfrank** [comment]
oh, no, i found another bug ... 
reproduce steps:
- recover account, logout
- remove account
- recover the same account again, UI will stuck here 🤷‍♂️:
image 

should i fix it in this PR? i believe it's not relate to my PR, but maybe it's brought by the rebasing of status-go  @pavloburykh

**Comment by pavloburykh** [comment]
> should i fix it in this PR? i believe it's not relate to my PR, but maybe it's brought by the rebasing of status-go @pavloburykh

@qfrank for sure it is brought by status-go rebasing. We need to fix it in this PR as we cannot introduce broken status-go to our develop. Thank you for catching those issues.

**Comment by qfrank** [comment]
> > should i fix it in this PR? i believe it's not relate to my PR, but maybe it's brought by the rebasing of status-go @pavloburykh
> 
> @qfrank for sure it is brought by status-go rebasing. We need to fix it in this PR as we cannot introduce broken status-go to our develop. Thank you for catching those issues.

should be fixed now 🙂

**Comment by pavloburykh** [comment]
@qfrank thank you for the fixes. 

Unfortunately there is a new issue

### ISSUE 3 User gets stucked on Sign in by syncing screen after scanning sync QR code

**Steps:** 

1. Device 1 generates sync QR
2. Device 2 scans QR
3. Observe the result

**Actual result:** Device 2 gets stucked on Sign in by syncing screen. Unable to perform syncing. 

[logs (9).zip](https://github.com/status-im/status-mobile/files/12618400/logs.9.zip)

https://github.com/status-im/status-mobile/assets/97245802/46606c19-4f73-4d27-a6be-62f80a9f0304

**Comment by qfrank** [comment]
> @qfrank thank you for the fixes.
> 
> Unfortunately there is a new issue
> 
> ### ISSUE 3 User gets stucked on Sign in by syncing screen after scanning sync QR code
> **Steps:**
> 
> 1. Device 1 generates sync QR
> 2. Device 2 scans QR
> 3. Observe the result
> 
> **Actual result:** Device 2 gets stucked on Sign in by syncing screen. Unable to perform syncing.
> 
> [logs (9).zip](https://github.com/status-im/status-mobile/files/12618400/logs.9.zip)
> 
>  scan_failed.mp4

the fix coming soon

**Comment by pavloburykh** [comment]
Thank you for the fixes. 

Unfortunately [ISSUE 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638) is still reproducible

https://github.com/status-im/status-mobile/assets/97245802/88b54d71-47b1-470d-9119-cd173546e5a1

Also there is a new issue

### ISSUE 4 Display name is not shown on Profile page after sync

**Steps:** 

1. Device 1 generates sync QR
2. Device 2 scans QR
3. See if display name is displayed on Profile page for Device 2 when sync is completed

**Actual result:** name field is empty

[Status-debug-logs (58).zip](https://github.com/status-im/status-mobile/files/12620155/Status-debug-logs.58.zip)

photo_2023-09-15 17 18 07


https://github.com/status-im/status-mobile/assets/97245802/3df99f3d-523c-4534-8cb4-5b14d06d928f


I really don't want to block this PR from merging anymore, but we need to fix ISSUEs 1, 2, 3 in followups after merge as they seem to be PR related. 


I am waiting for e2e run, once it is finished I'll ping you up

**Comment by qfrank** [comment]
> Thank you for the fixes.
> 
> Unfortunately [ISSUE 2](https://github.com/status-im/status-mobile/pull/17127#issuecomment-1717686638) is still reproducible
> 
>  telegram-cloud-document-2-5197696690643023345.mp4 
> Also there is a new issue
> 
> ### ISSUE 4 Display name is not shown on Profile page after sync
> **Steps:**
> 
> 1. Device 1 generates sync QR
> 2. Device 2 scans QR
> 3. See if display name is displayed on Profile page for Device 2 when sync is completed
> 
> **Actual result:** name field is empty
> 
> [Status-debug-logs (58).zip](https://github.com/status-im/status-mobile/files/12620155/Status-debug-logs.58.zip)
> 
> photo_2023-09-15 17 18 07
> 
>  telegram-cloud-document-2-5197696690643023322.mp4 
> I really don't want to block this PR from merging anymore, but we need to fix ISSUEs 1, 2, 3 in followups after merge as they seem to be PR related.
> 
> I am waiting for e2e run, once it is finished I'll ping you up

i'll take rest issues as follow up, let's merge now 🙂

**Comment by pavloburykh** [comment]
Thank you @qfrank 
E2E failures are not PR related. Ready to be merged.

**Pull request merged**
